src/pybind/mgr/cephadm/services/osd.py

```diff
@@ -656,25 +656,25 @@ def __init__(self):
         super().__init__()
 
     def as_osd_ids(self) -> List[int]:
-        return [osd.osd_id for osd in self]
+        return [osd.osd_id for osd in self.copy()]
```

Comment:
Is not this a smell of a potencial concurrency issue? The Runtime error usually happens when the iteration loop tries to remove or add items to the collection, but in this case it means that there's concurrent access to this object. Copying objects indiscriminately issues is a dangerous approach to solve concurrency issues. Proper locking should be ideal in this case.

Additionally, I cannot see how a `set` can be considered a queue, when sets are basically unordered collections (the opposite what a queue is)...

What about using a Python `Queue` with an `OrderedSet` container ([here you have a recipe](https://stackoverflow.com/questions/16506429/check-if-element-is-already-in-a-queue))?

Reply:
> Is not this a smell of a potencial concurrency issue? The Runtime error usually happens when the iteration loop tries to remove or add items to the collection, but in this case it means that there's concurrent access to this object. Copying objects indiscriminately is a dangerous approach to solve concurrency issues. Proper locking should be ideal in this case.

Agreed.

> Additionally, I cannot see how a `set` can be considered a queue, when sets are basically unordered collections (the opposite of what a queue is)...

Agreed. The naming `OSDQueue` probably suggests a ordered list of items. In this scenario however it's not as important if the items are ordered.




> What about using a Python `Queue` with an `OrderedSet` container ([here you have a recipe](https://stackoverflow.com/questions/16506429/check-if-element-is-already-in-a-queue))?

Sounds like a good solution. Let me see if I can adopt this approach here without breaking too much.

Reply:
See: c508c3654e47446bdb535845bb3cfc10f48ff43e